TypeScript error in adapt/pm.ts

## Error outputs

```
src/ts/adapt/pm.ts(140,3): error TS2376: A 'super' call must be the first statement in the constructor when a class contains initialized properties or has parameter properties.
src/ts/adapt/pm.ts(150,52): error TS17009: 'super' must be called before accessing 'this' in the constructor of a derived class.
src/ts/adapt/pm.ts(169,3): error TS2416: Property 'createInstance' in type 'PageMaster' is not assignable to the same property in base type 'PageBox'.
  Type '(parentInstance: any) => PageMasterInstance' is not assignable to type '(parentInstance: PageBoxInstance) => PageBoxInstance'.
    Type 'PageMasterInstance' is not assignable to type 'PageBoxInstance'.
      Property 'boxSpecificEnabled' is private in type 'PageBoxInstance' but not in type 'PageMasterInstance'.
src/ts/adapt/pm.ts(170,51): error TS2345: Argument of type 'this' is not assignable to parameter of type 'PageBox'.
  Type 'PageMaster' is not assignable to type 'PageBox'.
    Types of property 'createInstance' are incompatible.
      Type '(parentInstance: any) => PageMasterInstance' is not assignable to type '(parentInstance: PageBoxInstance) => PageBoxInstance'.
src/ts/adapt/pm.ts(176,3): error TS2416: Property 'clone' in type 'PageMaster' is not assignable to the same property in base type 'PageBox'.
  Type '(param: any) => PageMaster' is not assignable to type '(param: { parent: PageBox; pseudoName: string; }) => PageBox'.
    Type 'PageMaster' is not assignable to type 'PageBox'.
      Types of property 'createInstance' are incompatible.
        Type '(parentInstance: any) => PageMasterInstance' is not assignable to type '(parentInstance: PageBoxInstance) => PageBoxInstance'.
src/ts/adapt/pm.ts(183,24): error TS2345: Argument of type 'PageMaster' is not assignable to parameter of type 'PageBox'.
src/ts/adapt/pm.ts(184,24): error TS2345: Argument of type 'PageMaster' is not assignable to parameter of type 'PageBox'.
src/ts/adapt/pm.ts(809,21): error TS2352: Conversion of type 'Context' to type 'InstanceHolder' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property 'registerInstance' is missing in type 'Context'.
src/ts/adapt/pm.ts(1315,6): error TS2352: Conversion of type 'PageBoxInstance[]' to type 'PageMasterInstance[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'PageBoxInstance' is not comparable to type 'PageMasterInstance'.
    Property 'adjustPageLayout' is missing in type 'PageBoxInstance'.
src/ts/adapt/pm.ts(1317,33): error TS2339: Property 'specificity' does not exist on type 'PageBox'.
src/ts/adapt/pm.ts(1317,57): error TS2339: Property 'specificity' does not exist on type 'PageBox'.
src/ts/adapt/pm.ts(1322,14): error TS2415: Class 'PageMasterInstance' incorrectly extends base class 'PageBoxInstance'.
  Property 'boxSpecificEnabled' is private in type 'PageBoxInstance' but not in type 'PageMasterInstance'.
src/ts/adapt/pm.ts(1359,14): error TS2415: Class 'PartitionInstance' incorrectly extends base class 'PageBoxInstance'.
  Property 'boxSpecificEnabled' is private in type 'PageBoxInstance' but not in type 'PartitionInstance'.
src/ts/adapt/pm.ts(1520,25): error TS2345: Argument of type 'PageMaster' is not assignable to parameter of type 'PageBox'.
```

## Errors depending on other files

* adapt/cssparse.ts #471
* adapt/cssstyler.ts #473
* adapt/cssvald.ts #472
* adapt/expr.ts #467
* adapt/layout.ts #475
* adapt/net.ts #470
* adapt/task.ts #466
* adapt/taskutiil.ts #468
* adapt/vgen.ts #474
* adapt/vtree.ts #480 
* adapt/xmldoc.ts #469
* vivliostyle/layoututil.ts #476
* vivliostyle/repetitiveelements.ts #478
* vivliostyle/selectors.ts #479
* vivliostyle/table.ts #477

--- a/src/ts/adapt/base.ts
+++ b/src/ts/adapt/base.ts
@@ -593,7 +593,7 @@ export function multiIndexArray<T> (arr: T[], key: (p1: T) => string | null): {[
  * Apply function to each value of the object
  * @param fn second parameter is the key
  */
-export function mapObj<P, R> (obj: {[key: string]: P}, fn: (p1: any, p2: string) => R): {[key: string]: R} {
+export function mapObj<P, R> (obj: {[key: string]: P}, fn: (p1: P, p2: string) => R): {[key: string]: R} {
       const res: {[key: string]: R} = {};
       for (const n in obj) {
         res[n] = fn(obj[n], n);

--- a/src/ts/adapt/base.ts
+++ b/src/ts/adapt/base.ts
@@ -51,12 +51,18 @@ export const stripFragmentAndQuery = (url: string): string => {
  * Base URL relative to which URLs of resources are resolved.
  */
 export let baseURL = window.location.href;
+export function setBaseURL(value: string) {
+  baseURL = value;
+}
 
 /**
  * Base URL relative to which URLs of resources such as validation.txt and
  * user-agent.css are resolved.
  */
 export let resourceBaseURL = window.location.href;
+export function setResourceBaseURL(value: string) {
+  resourceBaseURL = value;
+}
 
 /**
  * @param relURL relative URL

--- a/src/ts/adapt/csscasc.ts
+++ b/src/ts/adapt/csscasc.ts
@@ -2728,7 +2728,10 @@ export enum ParseState {
 /**
  * Cascade for base User Agent stylesheet.
  */
-export let uaBaseCascade: Cascade = null;
+export var uaBaseCascade: Cascade = null;
+export function setUABaseCascade(value: Cascade) {
+  uaBaseCascade = value;
+}
 
 //------------- parsing ------------
 export class CascadeParserHandler extends

--- a/src/ts/adapt/epub.ts
+++ b/src/ts/adapt/epub.ts
@@ -1948,7 +1948,7 @@ export class OPFView implements CustomRendererFactory {
     }
   }
 
-  isTOCVisible(autohide): boolean {
+  isTOCVisible(): boolean {
     return this.tocView && this.tocView.isTOCVisible();
   }
 }

--- a/src/ts/adapt/epub.ts
+++ b/src/ts/adapt/epub.ts
@@ -61,7 +61,8 @@ export class EPUBDocStore extends ops.OPSDocStore {
   documents: {[key: string]: Result<xmldocs.XMLDocHolder>} = {};
 
   constructor() {
-    super(this.makeDeobfuscatorFactory());
+    super(null);
+    this.fontDeobfuscator = this.makeDeobfuscatorFactory();
     this.plainXMLStore = xmldocs.newXMLDocStore();
     this.jsonStore = net.newJSONStore();
   }
@@ -163,7 +164,7 @@ export class EPUBDocStore extends ops.OPSDocStore {
   }
 
   addDocument(url: string, doc: Document) {
-    const frame = task.newFrame('EPUBDocStore.load');
+    const frame = task.newFrame<xmldocs.XMLDocHolder>('EPUBDocStore.load');
     const docURL = base.stripFragment(url);
     const r = this.documents[docURL] = this.parseOPSResource({
       status: 200,
@@ -180,14 +181,14 @@ export class EPUBDocStore extends ops.OPSDocStore {
   /**
    * @override
    */
-  load(url) {
+  load(url: string)  {
     const docURL = base.stripFragment(url);
     let r = this.documents[docURL];
     if (r) {
       return r.isPending() ? r : task.newResult(r.get());
     } else {
-      const frame = task.newFrame('EPUBDocStore.load');
-      r = super(
+      const frame = task.newFrame<xmldocs.XMLDocHolder>('EPUBDocStore.load');
+      r = super.load(
           docURL, true,
           `Failed to fetch a source document from ${docURL}`);
       r.then((xmldoc: xmldocs.XMLDocHolder) => {
@@ -385,7 +386,7 @@ export const readMetadata = (mroot: xmldocs.NodeList, prefixes: string|null): JS
   let order = 1;
 
   // List of metadata items.
-  const rawItems = mroot.childElements().forEachNonNull((node) => {
+  const rawItems = mroot.childElements().forEachNonNull((node: Element) => {
     if (node.localName == 'meta') {
       const p = resolveProperty((node as Element).getAttribute('property'));
       if (p) {
@@ -416,9 +417,9 @@ export const readMetadata = (mroot: xmldocs.NodeList, prefixes: string|null): JS
   // Items grouped by their target id.
   const rawItemsByTarget =
       base.multiIndexArray(rawItems, (rawItem) => rawItem.refines);
-  const makeMetadata = (map) =>
-      base.mapObj(map, (rawItemArr, itemName) => {
-        const result = rawItemArr.map(rawItem => {
+  const makeMetadata = (map: {[key: string]: any[]}): {[key: string]: any[]} =>
+      base.mapObj(map, (rawItemArr, itemName) =>
+        rawItemArr.map(rawItem => {
           const entry = {'v': rawItem.value, 'o': rawItem.order};
           if (rawItem.schema) {
             entry['s'] = rawItem.scheme;
@@ -449,18 +450,17 @@ export const readMetadata = (mroot: xmldocs.NodeList, prefixes: string|null): JS
             }
           }
           return entry;
-        });
-        return result;
-      });
+        })
+      );
   const metadata = makeMetadata(base.multiIndexArray(
       rawItems, (rawItem) => rawItem.refines ? null : rawItem.name));
   let lang = null;
   if (metadata[metaTerms.language]) {
     lang = metadata[metaTerms.language][0]['v'];
   }
-  const sortMetadata = (metadata) => {
+  const sortMetadata = (metadata: {[key: string]: any[]}) => {
     for (const term in metadata) {
-      const arr = (metadata[term] as Array);
+      const arr = metadata[term];
       arr.sort(getMetadataComparator(term, lang));
       for (let i = 0; i < arr.length; i++) {
         const r = arr[i]['r'];
@@ -598,7 +598,7 @@ export class OPFDoc {
 
   initWithXMLDoc(
       opfXML: xmldocs.XMLDocHolder, encXML: xmldocs.XMLDocHolder,
-      zipMetadata: JSON, manifestURL: string): Result {
+      zipMetadata: JSON, manifestURL: string): Result<any> {
     const self = this;
     this.opfXML = opfXML;
     this.encXML = encXML;
@@ -938,7 +938,7 @@ export {OPFViewItem};
 
 export class OPFView implements CustomRendererFactory {
   spineItems: OPFViewItem[] = [];
-  spineItemLoadingContinuations: Continuation[][] = [];
+  spineItemLoadingContinuations: Continuation<any>[][] = [];
   pref: any;
   clientLayout: any;
   counterStore: any;
@@ -1006,8 +1006,8 @@ export class OPFView implements CustomRendererFactory {
    * too (calling `renderSinglePage` recursively).
    */
   private renderSinglePage(viewItem: OPFViewItem, pos: LayoutPosition):
-      Result<OPFView.RenderSinglePageResult> {
-    const frame: Frame<OPFView.RenderSinglePageResult> =
+      Result<RenderSinglePageResult> {
+    const frame: Frame<RenderSinglePageResult> =
         task.newFrame('renderSinglePage');
     let page = this.makePage(viewItem, pos);
     const self = this;
@@ -1641,7 +1641,7 @@ export class OPFView implements CustomRendererFactory {
           sb.append(srcParam);
           sb.append('&type=');
           sb.append(typeParam);
-          for (let c = srcElem.firstChild; c; c = c.nextSibling) {
+          for (let c: Node = srcElem.firstChild; c; c = c.nextSibling) {
             if (c.nodeType == 1) {
               const ce = (c as Element);
               if (ce.localName == 'param' &&

--- a/src/ts/adapt/layout.ts
+++ b/src/ts/adapt/layout.ts
@@ -2663,7 +2663,7 @@ export class Column extends vtree.Container {
                     return;
                   }
                 }
-                const viewTag = nodeContext.viewNode.localName;
+                const viewTag = (nodeContext.viewNode as Element).localName;
                 if (mediaTags[viewTag]) {
                   // elements that have inherent content
                   // check if a forced break must occur before the block.
@@ -2811,7 +2811,7 @@ export class Column extends vtree.Container {
                 // Leading edge
                 breakAtTheEdge = breaks.resolveEffectiveBreakValue(
                     breakAtTheEdge, nodeContext.breakBefore);
-                const viewTag = nodeContext.viewNode.localName;
+                const viewTag = (nodeContext.viewNode as Element).localName;
                 if (mediaTags[viewTag]) {
                   // elements that have inherent content
                   // check if a forced break must occur before the block.

--- a/src/ts/adapt/ops.ts
+++ b/src/ts/adapt/ops.ts
@@ -38,7 +38,7 @@ import * as exprs from './expr';
 import * as font from './font';
 import * as geom from './geom';
 import * as layout from './layout';
-import {Response, XMLHttpRequestResponseType} from './net';
+import {Response, XMLHttpRequestResponseType, ResourceStore} from './net';
 import * as pm from './pm';
 import * as task from './task';
 import {Fetcher} from './taskutil';
@@ -61,7 +61,7 @@ export const uaStylesheetBaseFetcher: Fetcher<boolean> =
         const handler = new csscasc.CascadeParserHandler(
             null, null, null, null, null, validatorSet, true);
         handler.startStylesheet(cssparse.StylesheetFlavor.USER_AGENT);
-        csscasc.uaBaseCascade = handler.cascade;
+        csscasc.setUABaseCascade(handler.cascade);
         cssparse.parseStylesheetFromURL(url, handler, null, null)
             .thenFinish(frame);
       });
@@ -543,7 +543,7 @@ export class StyleInstance extends exprs.Context implements
     const pageFloatLayoutContext = column.pageFloatLayoutContext;
     const deferredFloats =
         pageFloatLayoutContext.getDeferredPageFloatContinuations();
-    const frame = task.newFrame('layoutDeferredPageFloats');
+    const frame = task.newFrame<boolean>('layoutDeferredPageFloats');
     let invalidated = false;
     let i = 0;
     frame
@@ -1584,7 +1584,7 @@ export const parseOPSResource = (response: Response, store: xmldocs.XMLDocStore)
                                     task.Result<xmldocs.XMLDocHolder> =>
     (store as OPSDocStore).parseOPSResource(response);
 
-export class OPSDocStore extends xmldocs.XMLDocStore {
+export class OPSDocStore extends ResourceStore<xmldocs.XMLDocHolder> {
   styleByKey: {[key: string]: Style} = {};
   styleFetcherByKey: {[key: string]: Fetcher<Style>} = {};
   styleByDocURL: {[key: string]: Style} = {};
@@ -1607,7 +1607,7 @@ export class OPSDocStore extends xmldocs.XMLDocStore {
     this.setStyleSheets(authorStyleSheets as any, userStyleSheets as any);
     const userAgentXML =
         base.resolveURL('user-agent.xml', base.resourceBaseURL);
-    const frame = task.newFrame('OPSDocStore.init');
+    const frame = task.newFrame<boolean>('OPSDocStore.init');
     const self = this;
     cssvalid.loadValidatorSet().then((validatorSet) => {
       self.validatorSet = validatorSet;
@@ -1727,7 +1727,7 @@ export class OPSDocStore extends xmldocs.XMLDocStore {
       });
       const head = xmldoc.head;
       if (head) {
-        for (let c = head.firstChild; c; c = c.nextSibling) {
+        for (let c: Node = head.firstChild; c; c = c.nextSibling) {
           if (c.nodeType != 1) {
             continue;
           }

--- a/src/ts/adapt/pm.ts
+++ b/src/ts/adapt/pm.ts
@@ -34,19 +34,27 @@ export let keyCount: number = 1;
  * Represent an at-rule which creates a page-level CSS box (page-master,
  * partition, and partition-group).
  */
-export class PageBox {
+export abstract class PageBox<I extends PageBoxInstance = PageBoxInstance> {
   // styles specified in the at-rule
-  specified: any = ({} as csscasc.ElementStyle);
-  children: any = ([] as PageBox[]);
+  specified: csscasc.ElementStyle = {};
+  children: PageBox[] = [];
   pageMaster: PageMaster = null;
   index: number = 0;
   key: any;
 
+  protected _scope: exprs.LexicalScope;
+
+  get scope(): exprs.LexicalScope {
+    return this._scope;
+  }
+
   constructor(
-      public readonly scope: exprs.LexicalScope,
+      scope: exprs.LexicalScope,
       public readonly name: string|null,
       public readonly pseudoName: string|null,
-      public readonly classes: string[], public readonly parent: PageBox) {
+      public readonly classes: string[],
+      public readonly parent: PageBox) {
+    this._scope = scope;
     this.key = `p${keyCount++}`;
     if (parent) {
       this.index = parent.children.length;
@@ -54,7 +62,7 @@ export class PageBox {
     }
   }
 
-  createInstance(parentInstance: PageBoxInstance): PageBoxInstance {
+  createInstance(parentInstance: PageBoxInstance): I {
     throw new Error('E_UNEXPECTED_CALL');
   }
 
@@ -63,8 +71,7 @@ export class PageBox {
    * @param param parent: The parent of the cloned PageBox. pseudoName: Assign
    *     this value as the pseudoName of the cloned PageBox.
    */
-  clone(param: {parent: PageBox|undefined, pseudoName: string|undefined}):
-      PageBox {
+  clone(param: {parent?: PageBox, pseudoName?: string}): PageBox<I> {
     throw new Error('E_UNEXPECTED_CALL');
   }
 
@@ -96,7 +103,7 @@ export class PageBox {
 /**
  * Parent of all page masters
  */
-export class RootPageBox extends PageBox {
+export class RootPageBox extends PageBox<RootPageBoxInstance> {
   constructor(scope: exprs.LexicalScope) {
     super(scope, null, null, [], null);
     this.specified['width'] = new csscasc.CascadeValue(css.fullWidth, 0);
@@ -133,7 +140,7 @@ export class PageMasterScope extends exprs.LexicalScope {
 /**
  * Represent a page-master rule
  */
-export class PageMaster extends PageBox {
+export class PageMaster extends PageBox<PageMasterInstance> {
   pageMaster: any;
   keyMap: {[key: string]: string} = {};
 
@@ -142,14 +149,11 @@ export class PageMaster extends PageBox {
       classes: string[], parent: RootPageBox,
       public readonly condition: exprs.Val,
       public readonly specificity: number) {
-    let pageMasterScope;
-    if (scope instanceof PageMasterScope) {
-      // if PageMasterScope object is passed, use (share) it.
-      pageMasterScope = scope;
-    } else {
-      pageMasterScope = new PageMasterScope(scope, this);
+    super(scope, name, pseudoName, classes, parent);
+    // if PageMasterScope object is passed, use (share) it.
+    if (!(scope instanceof PageMasterScope)) {
+      this._scope = new PageMasterScope(scope, this);
     }
-    super(pageMasterScope, name, pseudoName, classes, parent);
     this.pageMaster = this;
     this.specified['width'] = new csscasc.CascadeValue(css.fullWidth, 0);
     this.specified['height'] = new csscasc.CascadeValue(css.fullHeight, 0);
@@ -201,7 +205,7 @@ export class PageMaster extends PageBox {
 /**
  * Represent a partition-group rule
  */
-export class PartitionGroup extends PageBox {
+export class PartitionGroup extends PageBox<PartitionGroupInstance> {
   pageMaster: any;
 
   constructor(
@@ -238,7 +242,7 @@ export class PartitionGroup extends PageBox {
 /**
  * Represent a partition rule
  */
-export class Partition extends PageBox {
+export class Partition extends PageBox<PartitionInstance> {
   pageMaster: any;
 
   constructor(
@@ -347,7 +351,7 @@ export const toExprBool =
       return val.toExpr(scope, scope.zero);
     };
 
-export interface InstanceHolder {
+export interface InstanceHolder extends exprs.Context {
   registerInstance(key: string, instance: PageBoxInstance): void;
 
   /**
@@ -543,9 +547,9 @@ export class PageBoxInstance {
     style['enabled'] = new css.Expr(enabled);
   }
 
-  private boxSpecificEnabled(enabled: exprs.Val): exprs.Val {return enabled}
+  protected boxSpecificEnabled(enabled: exprs.Val): exprs.Val {return enabled}
 
-      protected initHorizontal(): void {
+  protected initHorizontal(): void {
     const scope = this.pageBox.scope;
     const style = this.style;
     const parentWidth = this.parentInstance ?
@@ -1313,9 +1317,9 @@ export class RootPageBoxInstance extends PageBoxInstance {
     // Sort page masters using order and specificity.
     const pageMasters = this.children;
     (pageMasters as PageMasterInstance[])
-        .sort(
-            (a, b) => b.pageBox.specificity - a.pageBox.specificity ||
-                a.pageBox.index - b.pageBox.index);
+      .sort((a, b) =>
+        (b.pageBox as any).specificity - (a.pageBox as any).specificity // probably cause NaN
+        || a.pageBox.index - b.pageBox.index);
   }
 }
 

--- a/src/ts/adapt/pm.ts
+++ b/src/ts/adapt/pm.ts
@@ -34,7 +34,7 @@ export let keyCount: number = 1;
  * Represent an at-rule which creates a page-level CSS box (page-master,
  * partition, and partition-group).
  */
-export abstract class PageBox<I extends PageBoxInstance = PageBoxInstance> {
+export abstract class PageBox<I extends PageBoxInstance = PageBoxInstance<any>> {
   // styles specified in the at-rule
   specified: csscasc.ElementStyle = {};
   children: PageBox[] = [];
@@ -62,7 +62,7 @@ export abstract class PageBox<I extends PageBoxInstance = PageBoxInstance> {
     }
   }
 
-  createInstance(parentInstance: PageBoxInstance): I {
+  createInstance(parentInstance: PageBoxInstance): PageBoxInstance {
     throw new Error('E_UNEXPECTED_CALL');
   }
 
@@ -140,7 +140,7 @@ export class PageMasterScope extends exprs.LexicalScope {
 /**
  * Represent a page-master rule
  */
-export class PageMaster extends PageBox<PageMasterInstance> {
+export class PageMaster<I extends PageMasterInstance = PageMasterInstance<any>> extends PageBox<I> {
   pageMaster: any;
   keyMap: {[key: string]: string} = {};
 
@@ -170,7 +170,7 @@ export class PageMaster extends PageBox<PageMasterInstance> {
   /**
    * @override
    */
-  createInstance(parentInstance): PageMasterInstance {
+  createInstance(parentInstance): PageBoxInstance {
     return new PageMasterInstance(parentInstance, this);
   }
 
@@ -242,7 +242,7 @@ export class PartitionGroup extends PageBox<PartitionGroupInstance> {
 /**
  * Represent a partition rule
  */
-export class Partition extends PageBox<PartitionInstance> {
+export class Partition<I extends PartitionInstance = PartitionInstance> extends PageBox<I> {
   pageMaster: any;
 
   constructor(
@@ -258,7 +258,7 @@ export class Partition extends PageBox<PartitionInstance> {
   /**
    * @override
    */
-  createInstance(parentInstance) {
+  createInstance(parentInstance): PageBoxInstance {
     return new PartitionInstance(parentInstance, this);
   }
 
@@ -360,12 +360,12 @@ export interface InstanceHolder extends exprs.Context {
   lookupInstance(key: string): PageBoxInstance;
 }
 
-export class PageBoxInstance {
+export class PageBoxInstance<P extends PageBox = PageBox<any>> {
   /**
    * cascaded styles, geometric ones converted to css.Expr
    */
-  protected cascaded: any = ({} as csscasc.ElementStyle);
-  protected style: any = ({} as {[key: string]: css.Val});
+  protected cascaded: csscasc.ElementStyle = {};
+  style: {[key: string]: css.Val} = {};
   private autoWidth: exprs.Native = null;
   private autoHeight: exprs.Native = null;
   children: PageBoxInstance[] = [];
@@ -384,7 +384,7 @@ export class PageBoxInstance {
 
   constructor(
       public readonly parentInstance: PageBoxInstance,
-      public readonly pageBox: PageBox) {
+      public readonly pageBox: P) {
     if (parentInstance) {
       parentInstance.children.push(this);
     }
@@ -1302,7 +1302,7 @@ export const delayedProperties = ['transform', 'transform-origin'];
 
 export const userAgentPageMasterPseudo = 'background-host';
 
-export class RootPageBoxInstance extends PageBoxInstance {
+export class RootPageBoxInstance extends PageBoxInstance<RootPageBox> {
   constructor(pageBox: RootPageBox) {
     super(null, pageBox);
   }
@@ -1323,10 +1323,10 @@ export class RootPageBoxInstance extends PageBoxInstance {
   }
 }
 
-export class PageMasterInstance extends PageBoxInstance {
+export class PageMasterInstance<P extends PageMaster = PageMaster<PageMasterInstance<any>>> extends PageBoxInstance<P> {
   pageMasterInstance: any;
 
-  constructor(parentInstance: PageBoxInstance, pageBox: PageBox) {
+  constructor(parentInstance: PageBoxInstance, pageBox: P) {
     super(parentInstance, pageBox);
     this.pageMasterInstance = this;
   }
@@ -1351,7 +1351,7 @@ export class PageMasterInstance extends PageBoxInstance {
       clientLayout: vtree.ClientLayout) {}
 }
 
-export class PartitionGroupInstance extends PageBoxInstance {
+export class PartitionGroupInstance extends PageBoxInstance<PartitionGroup> {
   pageMasterInstance: any;
 
   constructor(parentInstance: PageBoxInstance, pageBox: PageBox) {
@@ -1360,10 +1360,10 @@ export class PartitionGroupInstance extends PageBoxInstance {
   }
 }
 
-export class PartitionInstance extends PageBoxInstance {
+export class PartitionInstance<P extends Partition = Partition<PartitionInstance<any>>> extends PageBoxInstance<P> {
   pageMasterInstance: any;
 
-  constructor(parentInstance: PageBoxInstance, pageBox: PageBox) {
+  constructor(parentInstance: PageBoxInstance, pageBox: P) {
     super(parentInstance, pageBox);
     this.pageMasterInstance = parentInstance.pageMasterInstance;
   }

--- a/src/ts/adapt/toc.ts
+++ b/src/ts/adapt/toc.ts
@@ -58,7 +58,7 @@ export class TOCView implements vgen.CustomRendererFactory {
     if (depth-- == 0) {
       return;
     }
-    for (let c = elem.firstChild; c; c = c.nextSibling) {
+    for (let c: Node = elem.firstChild; c; c = c.nextSibling) {
       if (c.nodeType == 1) {
         const e = (c as Element);
         if (base.getCSSProperty(e, 'height', 'auto') != 'auto') {
@@ -180,7 +180,7 @@ export const toggleNodeExpansion = (evt: Event) => {
   const open = elem.textContent == bulletClosed;
   elem.textContent = open ? bulletOpen : bulletClosed;
   const tocNodeElem = (elem.parentNode as Element);
-  let c = tocNodeElem.firstChild;
+  let c: Node = tocNodeElem.firstChild;
   while (c) {
     if (c.nodeType != 1) {
       c = c.nextSibling;

--- a/src/ts/adapt/viewer.ts
+++ b/src/ts/adapt/viewer.ts
@@ -105,8 +105,8 @@ export class Viewer {
   pageSizes: {width: number, height: number}[];
 
   // force relayout
-  viewport: any;
-  opfView: any;
+  viewport: vgen.Viewport | null;
+  opfView: epub.OPFView;
 
   constructor(
       public readonly window: Window,
@@ -376,13 +376,12 @@ export class Viewer {
 
     // for backward compatibility
     if (typeof command['userAgentRootURL'] == 'string') {
-      base.baseURL =
-          command['userAgentRootURL'].replace(/resources\/?$/, '');
-      base.resourceBaseURL = command['userAgentRootURL'];
+      base.setBaseURL(command['userAgentRootURL'].replace(/resources\/?$/, ''));
+      base.setResourceBaseURL(command['userAgentRootURL']);
     }
     if (typeof command['rootURL'] == 'string') {
-      base.baseURL = command['rootURL'];
-      base.resourceBaseURL = `${base.baseURL}resources/`;
+      base.setBaseURL(command['rootURL']);
+      base.setResourceBaseURL(`${base.baseURL}resources/`);
     }
     if (typeof command['pageViewMode'] == 'string' &&
         command['pageViewMode'] !== this.pageViewMode) {
@@ -649,7 +648,7 @@ export class Viewer {
    * @param sync If true, get the necessary page synchronously (not waiting
    *     another rendering task)
    */
-  private showCurrent(page: Page, sync?: boolean): task.Result {
+  private showCurrent(page: Page, sync?: boolean): task.Result<null> {
     this.needRefresh = false;
     this.removePageListeners();
     const self = this;
@@ -757,11 +756,11 @@ export class Viewer {
     const self = this;
     this.setReadyState(constants.ReadyState.LOADING);
     this.cancelRenderingTask();
-    const task = task.currentTask().getScheduler().run(
+    const resizeTask = task.currentTask().getScheduler().run(
         () => task.handle(
             'resize',
             (frame) => {
-              self.renderTask = task;
+              self.renderTask = resizeTask;
               profile.profiler.registerStartTiming(
                   'render (resize)');
               self.reset();
@@ -785,7 +784,7 @@ export class Viewer {
                         self.opfView.renderAllPages() :
                         task.newResult(null);
                     r.then(() => {
-                      if (self.renderTask === task) {
+                      if (self.renderTask === resizeTask) {
                         self.renderTask = null;
                       }
                       profile.profiler.registerEndTiming(

--- a/src/ts/vivliostyle/column.ts
+++ b/src/ts/vivliostyle/column.ts
@@ -61,10 +61,7 @@ function setBlockSize(container: Container, size) {
   }
 }
 
-/**
- * @abstract
- */
-export class ColumnBalancer {
+export abstract class ColumnBalancer {
   originalContainerBlockSize: any;
 
   constructor(
@@ -119,21 +116,11 @@ export class ColumnBalancer {
 
   protected preBalance(layoutResult: ColumnLayoutResult) {}
 
-  /**
-   * @abstract
-   */
-  protected calculatePenalty(layoutResult: ColumnLayoutResult): number {}
+  protected abstract calculatePenalty(layoutResult: ColumnLayoutResult): number;
 
-  /**
-   * @abstract
-   */
-  protected hasNextCandidate(candidates: ColumnBalancingTrialResult[]):
-      boolean {}
+  protected abstract hasNextCandidate(candidates: ColumnBalancingTrialResult[]): boolean;
 
-  /**
-   * @abstract
-   */
-  protected updateCondition(candidates: ColumnBalancingTrialResult[]) {}
+  protected abstract updateCondition(candidates: ColumnBalancingTrialResult[]): void;
 
   protected postBalance() {
     setBlockSize(this.layoutContainer, this.originalContainerBlockSize);

--- a/src/ts/vivliostyle/page.ts
+++ b/src/ts/vivliostyle/page.ts
@@ -668,7 +668,7 @@ export const marginBoxesKey: string = '_marginBoxes';
  * Represent a page master generated for @page rules
  * @param style Cascaded style for @page rules
  */
-export class PageRuleMaster extends pm.PageMaster {
+export class PageRuleMaster extends pm.PageMaster<PageRuleMasterInstance> {
   private bodyPartitionKey: any;
   private pageMarginBoxes: any =
       ({} as {[key: string]: PageMarginBoxPartition});
@@ -728,7 +728,7 @@ export class PageRuleMaster extends pm.PageMaster {
  * Represent a partition placed in a PageRuleMaster
  * @param style Cascaded style for @page rules
  */
-export class PageRulePartition extends pm.Partition {
+export class PageRulePartition extends pm.Partition<PageRulePartitionInstance> {
   constructor(
       scope: exprs.LexicalScope, parent: PageRuleMaster,
       style: csscasc.ElementStyle, public readonly pageSize: PageSize) {
@@ -760,7 +760,7 @@ export class PageRulePartition extends pm.Partition {
   /**
    * @override
    */
-  createInstance(parentInstance): PageRulePartitionInstance {
+  createInstance(parentInstance): pm.PageBoxInstance {
     return new PageRulePartitionInstance(parentInstance, this);
   }
 }
@@ -768,7 +768,7 @@ export class PageRulePartition extends pm.Partition {
 /**
  * Represent a partition for a page-margin box
  */
-export class PageMarginBoxPartition extends pm.Partition {
+export class PageMarginBoxPartition extends pm.Partition<PageMarginBoxPartitionInstance> {
   constructor(
       scope: exprs.LexicalScope, parent: PageRuleMaster,
       public readonly marginBoxName: string, style: csscasc.ElementStyle) {
@@ -806,7 +806,7 @@ export class PageMarginBoxPartition extends pm.Partition {
   /**
    * @override
    */
-  createInstance(parentInstance): PageMarginBoxPartitionInstance {
+  createInstance(parentInstance): pm.PageBoxInstance {
     return new PageMarginBoxPartitionInstance(parentInstance, this);
   }
 }
@@ -823,7 +823,7 @@ type PageAreaDimension = {
 
 export {PageAreaDimension};
 
-export class PageRuleMasterInstance extends pm.PageMasterInstance {
+export class PageRuleMasterInstance extends pm.PageMasterInstance<PageRuleMaster> {
   pageAreaDimension: PageAreaDimension|null = null;
   pageMarginBoxInstances: {[key: string]: PageMarginBoxPartitionInstance} = {};
 
@@ -1215,7 +1215,7 @@ interface MarginBoxSizingParam {
  */
 class SingleBoxMarginBoxSizingParam implements MarginBoxSizingParam {
   private hasAutoSize_: boolean;
-  private size: {[keykey in keyof sizing.Size]: number} | null = null;
+  private size: {[key in sizing.Size]: number} | null = null;
 
   constructor(
       protected readonly container: vtree.Container,
@@ -1382,7 +1382,7 @@ class FixedSizeMarginBoxSizingParam extends SingleBoxMarginBoxSizingParam {
   }
 };
 
-export class PageRulePartitionInstance extends pm.PartitionInstance {
+export class PageRulePartitionInstance extends pm.PartitionInstance<PageRulePartition> {
   borderBoxWidth: exprs.Val = null;
   borderBoxHeight: exprs.Val = null;
   marginTop: exprs.Val = null;
@@ -1537,7 +1537,7 @@ export class PageRulePartitionInstance extends pm.PartitionInstance {
   }
 }
 
-export class PageMarginBoxPartitionInstance extends pm.PartitionInstance {
+export class PageMarginBoxPartitionInstance extends pm.PartitionInstance<PageMarginBoxPartition> {
   boxInfo: PageMarginBoxInformation;
   suppressEmptyBoxGeneration: any = true;
 
@@ -1556,7 +1556,7 @@ export class PageMarginBoxPartitionInstance extends pm.PartitionInstance {
    */
   prepareContainer(context, container, page, docFaces, clientLayout) {
     this.applyVerticalAlign(context, container.element);
-    super(
+    super.prepareContainer(
         context, container, page, docFaces, clientLayout);
   }
 
@@ -1688,7 +1688,7 @@ export class PageMarginBoxPartitionInstance extends pm.PartitionInstance {
       [borderInsideWidth, paddingInside, paddingOutside, borderOutsideWidth]
           .forEach((x) => {
             if (x) {
-              borderAndPadding += x.evaluate(context);
+              borderAndPadding += x.evaluate(context) as number;
             }
           });
       if (result.marginInside === null || result.marginOutside === null) {
@@ -1795,7 +1795,7 @@ export class PageMarginBoxPartitionInstance extends pm.PartitionInstance {
    */
   finishContainer(
       context, container, page, column, columnCount, clientLayout, docFaces) {
-    super(
+    super.finishContainer(
         context, container, page, column, columnCount, clientLayout,
         docFaces);
 
@@ -1982,7 +1982,7 @@ export class PageManager {
       }
     });
     const pageMasterInstance =
-        newPageMaster.createInstance(this.rootPageBoxInstance);
+        newPageMaster.createInstance(this.rootPageBoxInstance) as pm.PageMasterInstance;
 
     // Do the same initialization as in adapt.ops.StyleInstance.prototype.init
     pageMasterInstance.applyCascadeAndInit(

--- a/src/ts/vivliostyle/viewerapp.ts
+++ b/src/ts/vivliostyle/viewerapp.ts
@@ -196,7 +196,7 @@ export const callback = (msg) => {
           'data-vivliostyle-page-progression', pageProgression);
       viewer.viewportElement.setAttribute(
           'data-vivliostyle-spread-view', viewer.pref.spreadView);
-      window.addEventListener('keydown', (keydown as Function), false);
+      window.addEventListener('keydown', keydown, false);
 
       //        window.addEventListener("touchstart", /** @type {Function} */
       //        (.touch), false);
@@ -207,12 +207,10 @@ export const callback = (msg) => {
       document.body.setAttribute('data-vivliostyle-viewer-status', 'complete');
       const leftButton =
           document.getElementById('vivliostyle-page-navigation-left');
-      leftButton.addEventListener(
-          'click', (navigateToLeftPage as Function), false);
+      leftButton.addEventListener('click', navigateToLeftPage, false);
       const rightButton =
           document.getElementById('vivliostyle-page-navigation-right');
-      rightButton.addEventListener(
-          'click', (navigateToRightPage as Function), false);
+      rightButton.addEventListener('click', navigateToRightPage, false);
       [leftButton, rightButton].forEach((button) => {
         button.setAttribute('data-vivliostyle-ui-state', 'attention');
         window.setTimeout(() => {

--- a/typings/lib.d.ts
+++ b/typings/lib.d.ts
@@ -1,0 +1,6 @@
+interface Element {
+  // `setAttribute` seems to allow non-string values.
+  // https://github.com/Microsoft/TypeScript/issues/15368
+  setAttribute(qualifiedName: string, value: number): void;
+  setAttribute(qualifiedName: string, value: boolean): void;
+}

